[SPARK-36804][YARN] Support --verbose option in YARN mode

daugraph opened the pull request:
### What changes were proposed in this pull request?
Support --verbose option in YARN mode


### Why are the changes needed?
If we submit the spark application with the --verbose parameter in yarn mode, we will get the following exception:
```bash
Exception in thread "main" java.lang.IllegalArgumentException: Unknown/unsupported param List(--verbose)Exception in thread "main"
```
SparkSubmit invoke YarnClusterApplication with --verbose arguments, however ClientArguments used by YarnClusterApplication don't support that argument by now, as a result, IllegalArgumentException is thrown in ClientArguments. I think we can support --verbose in YARN mode to keep consistency with other module


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
After apply this patch, application will submit successful

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -66,6 +74,15 @@ private[spark] class ClientArguments(args: Array[String]) {
     }
+
+    if (verbose) {
+      logInfo("Client arguments for YARN application:")
```
I remember we already have such mode in `SparkSubmit`. Can we follow how it's done there? e.g.) https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L86-L88

Printing out all args would be enough I suspect.

daugraph replied:
Thanks for your review, your are right, but this is two different --verbose option. SparkSubmit will pass --verbose option to org.apache.spark.deploy.yarn.ClientArgument.scala, which can't handle the --verbose by now. I will add some detailed description of screenshots later.

daugraph replied:
we can also avoid throw Expection by remove --verbose option before pass arguments to org.apache.spark.deploy.yarn.ClientArgument.scala

daugraph replied:
BTW, print all args is enough indeed

HyukjinKwon commented:
Can you show how you tested this? Showing a self-reproducer with before/after console message would be helpful for users. cc @tgravescs @mridulm FYI

HyukjinKwon commented:
Also, please enable GitHub Actions in your forked repository (see also https://github.com/apache/spark/pull/34046/checks?check_run_id=3648181789)

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #143530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143530/testReport)** for PR 34046 at commit [`80b24bd`](https://github.com/apache/spark/commit/80b24bdb8a4dd7cf2b46563d4708f9abdff0e540).

SparkQA commented:
**[Test build #143530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143530/testReport)** for PR 34046 at commit [`80b24bd`](https://github.com/apache/spark/commit/80b24bdb8a4dd7cf2b46563d4708f9abdff0e540).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143530/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48038/

daugraph commented:
### Source code repository: 
```bash
https://github.com/apache/spark.git -r 4ea54e8672757c0dbe3dd57c81763afdffcbcc1b
```
### Submit script/config:
```bash
export SPARK_PRINT_LAUNCH_COMMAND="1"
export SPARK_PREPEND_CLASSES="1"
export HADOOP_CONF_DIR=/path/to/hadoop/conf
export SPARK_SUBMIT_OPTS="-Djava.security.krb5.conf=/etc/krb5.conf"

spark-submit \
--master yarn \
--deploy-mode cluster \
--verbose \
--conf spark.kerberos.keytab=/path/to/keytab/file \
--conf spark.kerberos.principal=user_principal \
--conf spark.yarn.queue=root.user_queue \
--conf spark.yarn.maxAppAttempts=1 \
--class com.example.Main \
target/examples-1.0-SNAPSHOT.jar
```
### Output
```bash
NOTE: SPARK_PREPEND_CLASSES is set, placing locally compiled Spark classes ahead of assembly.
Spark Command: /Library/Java/JavaVirtualMachines/jdk1.8.0_271.jdk/Contents/Home/bin/java -cp /Users/lijianmeng/github/spark/conf/:/Users/lijianmeng/github/spark/common/kvstore/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/network-common/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/network-shuffle/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/network-yarn/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/sketch/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/tags/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/common/unsafe/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/core/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/examples/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/graphx/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/launcher/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/mllib/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/repl/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/resource-managers/mesos/target/scala-2.12/classes:/Users/lijianmeng/github/spark/resource-managers/yarn/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/sql/catalyst/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/sql/core/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/sql/hive/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/sql/hive-thriftserver/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/streaming/target/scala-2.12/classes/:/Users/lijianmeng/github/spark/core/target/jars/*:/Users/lijianmeng/github/spark/mllib/target/jars/*:/Users/lijianmeng/github/spark/assembly/target/scala-2.12/jars/*:/path/to/hadoop/conf/ -Djava.security.krb5.conf=/etc/krb5.conf org.apache.spark.deploy.SparkSubmit --master yarn --deploy-mode cluster --conf spark.kerberos.keytab=/path/to/keytab/file --conf spark.yarn.maxAppAttempts=1  --conf spark.kerberos.principal=user_principal --conf spark.yarn.queue=root.user_queue --class com.example.Main --verbose target/examples-1.0-SNAPSHOT.jar
========================================
Using properties file: null
Parsed arguments:
  master                  yarn
  deployMode              cluster
  executorMemory          null
  executorCores           null
  totalExecutorCores      null
  propertiesFile          null
  driverMemory            null
  driverCores             null
  driverExtraClassPath    null
  driverExtraLibraryPath  null
  driverExtraJavaOptions  null
  supervise               false
  queue                   root.user_queue
  numExecutors            null
  files                   null
  pyFiles                 null
  archives                null
  mainClass               com.example.Main
  primaryResource         file:/Users/lijianmeng/bigdata/examples/target/examples-1.0-SNAPSHOT.jar
  name                    com.example.Main
  childArgs               []
  jars                    null
  packages                null
  packagesExclusions      null
  repositories            null
  verbose                 true

Spark properties used, including those specified through
 --conf and those from the properties file null:
  (spark.yarn.queue,root.user_queue)
  (spark.yarn.maxAppAttempts,1)
  (spark.kerberos.principal,user_principal)
  (spark.kerberos.keytab,/path/to/keytab/file)


21/09/23 13:17:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Main class:
org.apache.spark.deploy.yarn.YarnClusterApplication
Arguments:
--jar
file:/Users/lijianmeng/bigdata/examples/target/examples-1.0-SNAPSHOT.jar
--class
com.example.Main
--verbose
Spark config:
(spark.kerberos.keytab,/path/to/keytab/file)
(spark.yarn.queue,root.user_queue)
(spark.app.name,com.example.Main)
(spark.kerberos.principal,user_principal)
(spark.submit.pyFiles,)
(spark.submit.deployMode,cluster)
(spark.yarn.maxAppAttempts,1)
(spark.master,yarn)
Classpath elements:
file:/Users/lijianmeng/bigdata/examples/target/examples-1.0-SNAPSHOT.jar


Exception in thread "main" java.lang.IllegalArgumentException: Unknown/unsupported param List(--verbose)

Usage: org.apache.spark.deploy.yarn.Client [options]
Options:
  --jar JAR_PATH           Path to your application's JAR file (required in YARN cluster
                           mode)
  --class CLASS_NAME       Name of your application's main class (required)
  --primary-py-file        A main Python file
  --primary-r-file         A main R file
  --arg ARG                Argument to be passed to your application's main class.
                           Multiple invocations are possible, each will be passed in order.

	at org.apache.spark.deploy.yarn.ClientArguments.parseArgs(ClientArguments.scala:61)
	at org.apache.spark.deploy.yarn.ClientArguments.<init>(ClientArguments.scala:31)
	at org.apache.spark.deploy.yarn.YarnClusterApplication.start(Client.scala:1701)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:955)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1043)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1052)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
21/09/23 13:17:38 INFO ShutdownHookManager: Shutdown hook called
21/09/23 13:17:38 INFO ShutdownHookManager: Deleting directory /private/var/folders/q4/bs9293m948l_rt31tm932s7w0000gn/T/spark-37708240-12b1-4013-9765-b7cc2c4d486e
```

@HyukjinKwon This is self-reproducer console log, cloud you review it? thanks for your time.

daugraph commented:
> Also, please enable GitHub Actions in your forked repository (see also https://github.com/apache/spark/pull/34046/checks?check_run_id=3648181789)

I have enable All Github Actions in my repo

SparkQA commented:
**[Test build #143536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143536/testReport)** for PR 34046 at commit [`7629af9`](https://github.com/apache/spark/commit/7629af9f55c6df09ab346febecd0915e63ea36b9).

SparkQA commented:
**[Test build #143536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143536/testReport)** for PR 34046 at commit [`7629af9`](https://github.com/apache/spark/commit/7629af9f55c6df09ab346febecd0915e63ea36b9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143536/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48038/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48038/

SparkQA commented:
**[Test build #143537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143537/testReport)** for PR 34046 at commit [`af65d25`](https://github.com/apache/spark/commit/af65d25eccf93f85f2f03dad1d338813b621475b).

SparkQA commented:
**[Test build #143537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143537/testReport)** for PR 34046 at commit [`af65d25`](https://github.com/apache/spark/commit/af65d25eccf93f85f2f03dad1d338813b621475b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143537/

SparkQA commented:
**[Test build #143538 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143538/testReport)** for PR 34046 at commit [`9e1afc8`](https://github.com/apache/spark/commit/9e1afc8c9ed0122b13ac992c6047981568020af8).

Commit:
[SPARK-36804][YARN] Using the verbose parameter in yarn mode would cause application submission failure

### What changes were proposed in this pull request?
Support --verbose option in YARN mode

### Why are the changes needed?
If we submit the spark application with the --verbose parameter in yarn mode, we will get the following exception:
```bash
Exception in thread "main" java.lang.IllegalArgumentException: Unknown/unsupported param List(--verbose)Exception in thread "main"
```
SparkSubmit invoke YarnClusterApplication with --verbose arguments, however ClientArguments used by YarnClusterApplication don't support that argument by now.
as a result, IllegalArgumentException is thrown in ClientArguments. I think we can support --verbose in YARN mode to keep consistency with other module

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
After apply this patch, application will submit successful

SparkQA commented:
**[Test build #143539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143539/testReport)** for PR 34046 at commit [`9bacaf2`](https://github.com/apache/spark/commit/9bacaf25f31596c80c3cec55799ada5f325ec261).

SparkQA commented:
**[Test build #143538 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143538/testReport)** for PR 34046 at commit [`9e1afc8`](https://github.com/apache/spark/commit/9e1afc8c9ed0122b13ac992c6047981568020af8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143538/

SparkQA commented:
**[Test build #143539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143539/testReport)** for PR 34046 at commit [`9bacaf2`](https://github.com/apache/spark/commit/9bacaf25f31596c80c3cec55799ada5f325ec261).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48045/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143539/

SparkQA commented:
**[Test build #143540 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143540/testReport)** for PR 34046 at commit [`087d069`](https://github.com/apache/spark/commit/087d0697d21af66a86eefb379a5a71b6270425f8).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48047/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48048/

SparkQA commented:
**[Test build #143540 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143540/testReport)** for PR 34046 at commit [`087d069`](https://github.com/apache/spark/commit/087d0697d21af66a86eefb379a5a71b6270425f8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143540/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48045/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48047/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48047/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48045/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48049/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48048/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48048/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48046/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48046/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48049/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48049/

SparkQA commented:
**[Test build #143550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143550/testReport)** for PR 34046 at commit [`662a54d`](https://github.com/apache/spark/commit/662a54d709155f369ccd7aeb26ac40155c8c32cc).

SparkQA commented:
**[Test build #143551 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143551/testReport)** for PR 34046 at commit [`5a5348a`](https://github.com/apache/spark/commit/5a5348aaee0b0004406cdb2fec436590559d2caf).

SparkQA commented:
**[Test build #143550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143550/testReport)** for PR 34046 at commit [`662a54d`](https://github.com/apache/spark/commit/662a54d709155f369ccd7aeb26ac40155c8c32cc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143550/

SparkQA commented:
**[Test build #143551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143551/testReport)** for PR 34046 at commit [`5a5348a`](https://github.com/apache/spark/commit/5a5348aaee0b0004406cdb2fec436590559d2caf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48059/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143551/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48060/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48059/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48060/

HyukjinKwon reviewed (commented):
I haven't tested it by myself but looks good from a cusory look.

daugraph commented:
> I haven't tested it by myself but looks good from a cusory look.

Take your time.

daugraph commented:
Can one of the admins verify this patch? Thanks @HyukjinKwon @mojodna @marmbrus

daugraph commented:
What else can I do for this patch ? @HyukjinKwon @srowen @mridulm

srowen reviewed (commented):
How about logging at a level lower than info, and then letting users choose to see the lower-level logs? that seems more elegant than adding a verbose flag

HyukjinKwon commented:
actually this passes the flags from spark-summit CLI to here. so this isn't actually a new flag but making an existing flag work (see the example above https://github.com/apache/spark/pull/34046#issuecomment-925517526)

srowen commented:
Ah OK SGTM

HyukjinKwon commented:
The change looks good .. but hope somebody takes a closer look of the implementation .. it's been a while since I touched Spark summit related codes and Yarn related codes ...

HyukjinKwon commented:
cc @vanzin and @jerryshao too FYI

srowen commented:
Merged to master
